### DISCUSSION TITLE
React Markdown V9.0.0: "Cannot find module unist-util-visit-parents/do-not-use-color"

### DISCUSSION THREAD

**Discussion by molamphd** [remarkjs/remark#1247] [Q&A] [answered]
After updating React Markdown to the latest version (9.0.0) and running unit tests on the components that use the package, I encountered an error (example attached below).

The unit tests check if the component renders without crashing. Interestingly, the components using the package still work fine after making the suggested migration changes. However, the issue with the unit tests on the components using React Markdown failing and throwing an error persists.

Any help would be much appreciated. I have hit a bit of a dead end after trying the usual troubleshooting techniques associated with this type of error.

<img width="1089" alt="Screenshot 2023-10-20 at 08 30 29" src="https://github.com/remarkjs/remark/assets/60659482/04e2da7c-3a02-4d1d-a2b2-c9420ae7de01">

**Comment by wooorm** [comment, marked as answer]
Hi! The problem is that your tooling is quite out of date. I think you’re using Jest? Jest has several problems with modern ESM. I recommend using `vitest` if you want something similar to Jest, or `node:test` if you want a more minimal less magic setup, that’s what I use everywhere

**Reply by ChristianMurphy** [reply]
Adding on, moving away from Jest and other tools which do not support ESM would be ideal.
If you need a migration step between your current test tool and a modern one.
You could wrap jest with another tool like vite which does support ESM already, here is blog post with example integration: https://hung.dev/posts/jest-vite

**Comment by hkjels** [comment]
It also fails if you try to use this library from ClojureScript with shadow-cljs.

**Reply by wooorm** [reply]
Then report it with those tools. They are buggy.

**Reply by ChristianMurphy** [reply]
@hkjels https://github.com/thheller/shadow-cljs/blob/master/doc/esm.md, https://tonsky.me/blog/clojurescript-2/, and https://github.com/squint-cljs/cherry may interest you.

**Comment by amila-desilva** [comment]
I am quite confused how Jest is the root cause of the issue... 
My react tests break at `import rehypeHighlight from 'rehype-highlight'; `  with the same error:  _cannot find  unist-util-visit-parents/do-not-use-color_. I looked thru the node_modules/unist-util-visit-parents/lib/index.js and line 221 (as shown in the screenshot) does not exist in the module, which is also why when my react tests run, it says it cannot find the module. Can someone help me get clarity around this issue?


![image](https://github.com/remarkjs/remark/assets/81698709/59ec7d3d-b9c4-417b-a6f1-e2bb54fae80d)

**Reply by ChristianMurphy** [reply]
Understanding how modules work would be a good starting point 🙂 
recommended reading: https://nodejs.org/api/packages.html

> I am quite confused how Jest is the root cause of the issue

Jest is both a test tool and a build tool.
As a build tool it does not support standard exports.

> does not exist in the module, which is also why when my react tests run, it says it cannot find the module

It uses an export to switch the implementation based on your environment.
https://github.com/syntax-tree/unist-util-visit-parents/blob/58e20d8366363b8b6e22d2e9c82d2ded408ecf16/package.json#L34-L37

Jest does not support standard package exports
https://jestjs.io/docs/ecmascript-modules
https://github.com/jestjs/jest/issues/11167
https://github.com/jestjs/jest/issues/9430

As noted above

> I recommend using `vitest` if you want something similar to Jest, or `node:test` if you want a more minimal less magic setup, that’s what I use everywhere

Both of these have much better support for standard ESM and package exports.

**Reply by amila-desilva** [reply]
Ahhh yes!! This is something new I learned. Vitest helped resolve the issue. Thanks Christian!

**Reply by studioswong** [reply]
Thank you for the above detailed explanation @ChristianMurphy and @wooorm !

Other than migrating to `vitest`, are there any other work around to enable Jest to render modules that uses `react-markdown`? While vitest might provide better support for ESM and package exports, given the wide adoption of Jest, sometimes migrating to `vitest` is not an option ( esp if your codebase is heavily invested in Jest, it might seem a bit of an overkill to migrate to `vitest` just for this.) 

Thank you!

**Reply by remcohaszing** [reply]
If you insist on using Jest, you can send bug fixes to Jest. You would do the JavaScript community a great favor.

But I recommend switching to Vitest.

**Reply by studioswong** [reply]
I currently uses `babel-jest` and the `@babel/plugin-transform-modules-commonjs` plugin to enable Jest to support ESM modules, but unfortunately the above error persists - am I missing something? should this technically work if `babel` is involved to help transform ESM modules in Jest?

( P.S I personally am a fan of `vite.js` and open to Vitest, but am not sure if we should switch to Vitest in the context of our current project given that we are not using `vite.js` as a build system, moreover with the migration and refactoring efforts involved 😅 )

**Reply by ChristianMurphy** [reply]
> but unfortunately the above error persists - am I missing something?

I don't know it is possible today in Jest, period.
Jest fundamentally has issues with ESM https://github.com/orgs/remarkjs/discussions/1247#discussioncomment-8084664

You can try the babel plugin, it may or may not fix it.
That is something you need to experiment with yourself.
On thing that catches my eye, it does not follow node resolution rules by default https://babel.dev/docs/babel-plugin-transform-modules-commonjs#options you may need those rules applied to follow the conditional exports correctly.

**Reply by ChristianMurphy** [reply]
To reiterate @remcohaszing's comment and my own previous one.
The level of effort moving to vite and vitest is probably significantly less than the effort of trying to work around Jest's current limitations.

**Reply by studioswong** [reply]
Gotcha, thank you @ChristianMurphy and @remcohaszing !

**Reply by alankalb** [reply]
@studioswong If you are still stuck with using Jest, you can try adding `unist-util-visit-parents` to `moduleNameMapper` in your jest config. I have found that jest is unable to find some files within packages like `unist-util-visit-parents` and `vfile`.

```
moduleNameMapper: {
    ...,
    "unist-util-visit-parents/do-not-use-color": "<rootDir>/node_modules/unist-util-visit-parents/lib/color.js",
    "vfile/do-not-use-conditional-minpath": "<rootDir>/node_modules/vfile/lib/minpath.browser.js",
    "vfile/do-not-use-conditional-minproc": "<rootDir>/node_modules/vfile/lib/minproc.browser.js",
    "vfile/do-not-use-conditional-minurl": "<rootDir>/node_modules/vfile/lib/minurl.browser.js",
  },
```

But like others have said, we should all be switching to vitest to avoid this nonsense.

**Reply by Jeff-Tian** [reply]
Thank you! @alankalb , I followed your example and fixed my issue here with adding the following to `jest.config.js`:

```
moduleNameMapper: {
    // ...
    "unist-util-visit-parents/do-not-use-color":
      "<rootDir>/node_modules/unist-util-visit-parents/lib/color.js",
    "^#minpath$": "<rootDir>/node_modules/vfile/lib/minpath.browser.js",
    "^#minproc$": "<rootDir>/node_modules/vfile/lib/minproc.browser.js",
    "^#minurl$": "<rootDir>/node_modules/vfile/lib/minurl.browser.js",
}
```
